uarray/_uarray_dispatch.cxx

```diff
@@ -179,6 +185,261 @@ std::string backend_to_domain_string(PyObject * backend) {
+  }
+
+  static backend_options convert_backend_options(PyObject * input) {
+    backend_options output;
+    int coerce, only;
+    PyObject * py_backend;
+    if (!PyArg_ParseTuple(input, "Opp", &py_backend, &coerce, &only))
+      throw std::invalid_argument("");
+
+    if (py_backend != Py_None) {
+      output.backend = py_ref::ref(py_backend);
+    }
+    output.coerce = coerce;
+    output.only = only;
+
+    return output;
+  }
+
+  static py_ref convert_backend(PyObject * input) { return py_ref::ref(input); }
+
+  static local_backends convert_local_backends(PyObject * input) {
+    PyObject *py_skipped, *py_preferred;
+    if (!PyArg_ParseTuple(input, "OO", &py_skipped, &py_preferred))
+      throw std::invalid_argument("");
+
+    local_backends output;
+    output.skipped =
+        convert_iter<py_ref>(py_skipped, BackendState::convert_backend);
+    output.preferred = convert_iter<backend_options>(
+        py_preferred, BackendState::convert_backend_options);
+
+    return output;
+  }
+
+  static global_backends convert_global_backends(PyObject * input) {
+    PyObject *py_global, *py_registered;
+    if (!PyArg_ParseTuple(input, "OO", &py_global, &py_registered))
+      throw std::invalid_argument("");
+
+    global_backends output;
+    output.global = BackendState::convert_backend_options(py_global);
+    output.registered =
+        convert_iter<py_ref>(py_registered, BackendState::convert_backend);
+
+    return output;
+  }
+
+  static global_state_t convert_global_state(PyObject * input) {
+    return convert_dict<std::string, global_backends>(
+        input, BackendState::convert_domain,
+        BackendState::convert_global_backends);
+  }
+
+  static local_state_t convert_local_state(PyObject * input) {
+    return convert_dict<std::string, local_backends>(
+        input, BackendState::convert_domain,
+        BackendState::convert_local_backends);
+  }
+
+  static py_ref convert_py(py_ref input) { return input; }
+
+  static py_ref convert_py(bool input) { return py_bool(input); }
+
+  static py_ref convert_py(backend_options input) {
+    if (!input.backend) {
+      input.backend = py_ref::ref(Py_None);
+    }
+    py_ref output = py_make_tuple(
+        input.backend, py_bool(input.coerce), py_bool(input.only));
+    if (!output)
+      throw std::runtime_error("");
+    return output;
+  }
+
+  static py_ref convert_py(const std::string & input) {
+    py_ref output =
+        py_ref::steal(PyUnicode_FromStringAndSize(input.c_str(), input.size()));
+    if (!output)
+      throw std::runtime_error("");
+    return output;
+  }
+
+  template <typename T>
+  static py_ref convert_py(const std::vector<T> & input) {
+    py_ref output = py_ref::steal(PyList_New(input.size()));
+
+    if (!output)
+      throw std::runtime_error("");
+
+    for (size_t i = 0; i < input.size(); i++) {
+      py_ref element = convert_py(input[i]);
+      Py_INCREF(element.get());
+      PyList_SET_ITEM(output.get(), i, element.get());
```

Comment:
Nit: Instead of `INCREF`, better to transfer ownership with `release()`:
```suggestion
      PyList_SET_ITEM(output.get(), i, element.release());
```